### ISSUE TITLE
Wrong ES3/ES5 code when iterated identifier is reassigned in a 'for...of' statement

### ISSUE THREAD

**Issue by olydis** [Microsoft/TypeScript#5475]
There is an inconsistency regarding the emitted ES3/ES5 code:

```` TypeScript
var x = [3, 3, 3];
for (var i of x)
{
	x = [2, 2];
	console.log(i);
}
````

compiles to 

```` JavaScript
var x = [3, 3, 3];
for (var _i = 0; _i < x.length; _i++) {
    var i = x[_i];
    x = [2, 2];
    console.log(i);
}
````

printing

    3
    2

instead of (ES6 semantics)

    3
    3
    3

(Note that the code emitted also differs from the feature description: https://github.com/Microsoft/TypeScript/pull/2207#issue-59900754)

---
On the other hand, merely adding parentheses around the expression:

```` TypeScript
var x = [3, 3, 3];
for (var i of (x))
{
    x = [2, 2];
    console.log(i);
}
````

**correctly** compiles to

```` JavaScript
var x = [3, 3, 3];
for (var _a = 0, _b = (x); _a < _b.length; _a++) {
    var i = _b[_a];
    x = [2, 2];
    console.log(i);
}
````

giving the expected output.

---
In other words, it looks like there is some sort of faulty optimization applied: As soon as the expression to iterate over is a bare variable, it is not rescued to a new local (here: `_b`). As shown, this results in wrong semantics.

**Comment by DanielRosenwasser** [comment]
Thanks for the bug report! You can try out our nightlies by installing `npm install -g typescript@next` by ~11:30 tonight or tomorrow morning and you should get the fix.

**Comment by olydis** [comment]
Amazing, thank you!
